zerver/views/realm_emoji.py

```diff
@@ -26,24 +27,40 @@ def upload_emoji(request: HttpRequest, user_profile: UserProfile,
                  emoji_name: Text=REQ()) -> HttpResponse:
     check_valid_emoji_name(emoji_name)
     check_emoji_admin(user_profile)
+    if RealmEmoji.objects.filter(realm=user_profile.realm,
+                                 name=emoji_name,
+                                 deactivated=False).exists():
+        return json_error(_("A realm emoji with this name already exists."))
     if len(request.FILES) != 1:
         return json_error(_("You must upload exactly one file."))
     emoji_file = list(request.FILES.values())[0]
     if (settings.MAX_EMOJI_FILE_SIZE * 1024 * 1024) < emoji_file.size:
         return json_error(_("Uploaded file is larger than the allowed limit of %s MB") % (
             settings.MAX_EMOJI_FILE_SIZE))
-    emoji_file_name = get_emoji_file_name(emoji_file.name, emoji_name)
-    upload_emoji_image(emoji_file, emoji_file_name, user_profile)
+
+    realm_emoji = check_add_realm_emoji(user_profile.realm,
+                                        emoji_name,
+                                        author=user_profile)
+    emoji_file_name = get_emoji_file_name(emoji_file.name, realm_emoji.id)
+    emoji_uploaded_successfully = False
     try:
-        check_add_realm_emoji(user_profile.realm, emoji_name, emoji_file_name, author=user_profile)
-    except ValidationError as e:
-        return json_error(e.messages[0])
```

Comment:
This path will never be triggered so removed this.

Reply:
Isn't this what happens if the `validators` on `class RealmEmoji` fire?  Or is the point that is no longer possible, because we verify the character set earlier anyway.

Reply:
> Or is the point that is no longer possible, because we verify the character set earlier anyway.

Yeah, this will never happen because of pre-validation.